shell/frameworks/namcap/namcap.1
.TH namcap 1 "Jan 14, 2024" "namcap 3.5.2" "User Commands"
.SH NAME
namcap \- package analysis utility
.SH SYNOPSIS
\fBnamcap [options] <package|PKGBUILD> [package|PKGBUILD] ...
.SH DESCRIPTION
.PP
\fBnamcap\fP is a \fIpackage analysis\fP utility that looks for problems with Arch Linux packages or their PKGBUILD files.  It can apply rules to the file list, the files themselves, or individual PKGBUILD files.
.PP
Rules return lists of messages.  Each message can be one of three types: error, warning, or information (think of them as notes or comments).  Errors (designated by 'E:') are things that namcap is very sure are wrong and need to be fixed.  Warnings (designated by 'W:') are things that namcap thinks should be changed but if you know what you're doing then you can leave them.  Information (designated 'I:') are only shown when you use the info argument.  Information messages give information that might be helpful but isn't anything that needs changing.
.SH OPTIONS
.TP
\fB\-e\fR RULELIST, \fB\-\-exclude=\fRRULELIST
Do not run RULELIST rules on the package
.TP
.B "\-i, \-\-info"
display information messages
.TP
.B "\-L, \-\-list
return a list of valid rules and their descriptions
.TP
.B "\-m, \-\-machine\-readable"
displays easily parseable namcap tags instead of the normal human readable description; for example using non-fhs-man-page instead of "Non-FHS man page (%s) found. Use /usr/share/man instead". A full list of namcap tags along with their human readable descriptions can be found at /usr/share/namcap/tags.
.TP
\fB\-r\fR RULELIST, \fB\-\-rules=\fRRULELIST
only apply RULELIST rules to the package
.IP
RULELIST is a comma-separated list of rule names
.TP
.B "\-v, \-\-version"
print version and exit
.SH RULES
.TP
.B arrays
Verifies that array variables are actually arrays and not strings
.TP
.B badbackups
Verifies that backup entries don't begin with /
.TP
.B carch
Looks for references to i686 and i586 and tells you to use $CARCH instead
.TP
.B capsnames
Checks a PKGBUILD to verify that the package name has no upper case characters
.TP
.B capsnamespkg
Checks a package to verify that the package name has no upper case characters
.TP
.B checksums
Makes sure that a PKGBUILD includes valid checksums
.TP
.B depends
This module runs ldd on all executables, gets the link-level dependencies, finds the smallest subset of dependencies that cover the link-level dependencies, and compares that list to the depends of the package.  It returns messages in three cases: dependency detected and not included, dependency included but already satisfied, and dependency included and not needed.  These suggestions are just guidelines and all package builders should take this into account (i.e. you're smarter than namcap is)

Some cases where namcap fails are dlopen() and obscure links.  dlopen()'d libraries don't show up because they are loaded at run time: in the case of a program that loads plugins.  Obscure links are the cases where only a small portion of the package needs something to run; usually, the small portion won't be accessed unless that thing is installed (i.e. a java plugin for mozilla included in the java package).
.TP
.B directoryname
Checks the directory names in the package file list for standard directories (i.e. /etc, /usr/bin, /usr/lib).  All files in directories that are non-standard are returned
.TP
.B elffiles
Gives an error if it finds ELF (binary) files outside the standard paths (/bin, /usr/bin, etc.)
.TP
.B emptydir
Looks for directories in a package that don't contain any files
.TP
.B extravars
Verifies that extra variables start with an underscore
.TP
.B fhsinfopages
Verifies correct installation of info pages
.TP
.B fhsmanpages
Verifies correct installation of man pages
.TP
.B fileownership
Verifies that all files are owned by root:root and not some other user or group
.TP
.B gnomemime
Checks for generated GNOME mime files
.TP
.B hicoloricons
Checks whether the hicolor icon cache is updated when a package installs
files in /usr/share/icons/hicolor
.TP
.B infodirectory
Verifies that a package does not contain the info directory file (/usr/share/info/dir)
.TP
.B invalidstartdir
Any PKGBUILDs that refer to files in $startdir are incorrect
.TP
.B libtool
Warns if libtool (*.la) files are present
.TP
.B license
Verifies that the licenses variable has been filled in in a PKGBUILD.
.TP
.B licensepkg
Verifies that the licenses variable has been filled in in a package. For packages with a custom license it
checks whether the license file has been installed in
/usr/share/licenses/$pkgname/
.TP
.B perllocal
Searches for perllocal.pod.  perllocal.pod is a nasty file that's included during most perl module installations
.TP
.B permissions
Checks basic file and and directory permissions.  It returns warnings about world readable bits not being set, world writable bits being set, and world executable bit not being set for directories
.TP
.B rpath
Gives an error if a binary has RPATH set to something other than /usr/lib
.TP
.B runpath
Gives an error if a binary has RUNPATH set to something other than /usr/lib, /usr/lib32
.TP
.B scrollkeeper
Verifies that there aren't any scrollkeeper directories
.TP
.B sfurl
Warns if a PKGBUILD is downloading from a specific sourceforge mirror instead of the generic downloads.sourceforge.net
.TP
.B symlink
Finds out symbolic and hard links which do not point to a valid path in the package
.TP
.B tags
For package maintainers, this module verifies that we have remembered to add a Maintainer and Contributor comment.
.TP
.B url
Verifies that we have the url variable set in the PKGBUILD
.TP
.B urlpkg
Verifies that we have the url variable set in the package file
.SH EXAMPLES
.TP
.B namcap foo.pkg.tar.gz
apply all rules to foo.pkg.tar.gz
.TP
.B namcap -r depends,symlink foo.pkg.tar.gz
apply the depends and symlink rules to foo.pkg.tar.gz
.TP
.B namcap PKGBUILD
apply all PKGBUILD based rules to the file PKGBUILD
.TP
.B namcap --list
list all of the available rules
.SH COPYRIGHT
Copyright \(co 2003-2023 Namcap contributors, see AUTHORS for details.
.PP
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.fi


CHANGES
2024.02.4, released July 13th, 2024

	Important / security related fixes.

	Infrastructure: Ensure a custom tar program specified with
	make TAR=/path/to/custom/tar is used everwhere.

	Defconfig: Pine64 rockpro64: Build ethernet driver in kernel
	rather than as module.

	Updated/fixed packages: apache, composer, daq, fluent-bit,
	ghostscript, glibc, liblockfile, libmad, libopenssl, mesa3d,
	network-manager, ntp, octave, openssh, openvpn, php, python3,
	qt6base, raspberrypi-usbboot, redis, sentry-cli, systemd, tor,
	vlc, xmrig, znc

2024.02.3, released June 10th, 2024

	Important / security related fixes.

	Misc:
	- Fixed a number of compilation issues for hosts using GCC
          14.x, including for kconfig.

	Download:
	- Drop redundant --passive-ftp from default wget options and
	  migrate a number of packages using FTP upstreams to HTTP(S)
	  for compatibility with wget2 which does not have FTP support

	- Ensure downloads are correctly shared for host packages that
	  try to do so using <pkg>_DL_SUBDIR and ensure the correct
	  toplevel directory name is used in generated tarballs when
	  downloads are shared.

	Arch: Enable SSE2 support for x86 Pentium M variant.

	Defconfigs: Orangepi PC/R1, Raspberrypi zero2w, Sipeed
	licheepi nano, Wandboard: Fix compilation issues. PC x86-64
	bios/efi: add wireless regdb to fix kernel warning during
	boot, iMX.8 boards: Adjust FIT external data position to match
	imx-mkimage changes, fixing boot issues.

	Updated/fixed packages: am33x-cm, apcupsd,
	arm-trusted-firmware, binutils-bare-metal, bpftool, catch2,
	ccache, cjson, clamav, cog, coreutils, cryptodev-linux,
	dhcpcd, dmraid, dnsmasq, docker-cli, docker-compose,
	docker-engine, doom-wad, duma, elf2flt, ebtables, elfutils,
	evemu, file, fluidsynth, freeradius-client, freeradius-server,
	freerdp, fwts, gcc, gcc-bare-metal, git, gnuplot, go,
	gr-osmosdr, gst1-plugins-good, hidapi, imagemagick,
	intel-gmmlib, ivi-homescreen, janus-gateway, kexec, kmod,
	kodi, libarchive, libdrm, libedit, libffi, libglib2, libhtp,
	libjwt, libkrb5, libmanette, libmdbx, libmnl,
	libnetfilter_acct, libnetfilter_log, libnpth, libopenssl,
	libpng, libuhttpd, liburiparser, libutempter, libvncserver,
	libwpe, lightning, linux-firmware, linux-pam, lua-cffi,
	luajit, lvm2, lynx, mariadb, mesa3d, micropython, minisign,
	musl, mutt, ncftp, network-manager, newlib-bare-metal, nfacct,
	nfs-utils, octave, openjpeg, openocd, oprofile,
	perl-dbd-mysql, php, postgresql, pppd, privoxy, procps-ng,
	proftpd, proj, putty, python-ml-dtypes, python-pillow,
	python-rds-py, qemu, qt5base, redis-plus-plus, rtl8189fs,
	ruby, samba4, shadow, smcroute, speech, squid, sscep,
	suricata, swupdate, systemd, ti-sgx-km, tinycbor, tinyproxy,
	tor, tpm2-tss, ttyd, uboot-tools, ulogd, vala, weston,
	wpewebkit, x11vnc, zbar, zeromq, zfs, zip, zxing-cpp

	Issues resolved (http://bugs.uclibc.org):

        #15766: LINGMO_PACKAGE_WESTON_XWAYLAND -- 12.0.1 -- fails to build..
	#15973: imx8m bootloader image generation broken since update..
        #16075: Cannot use linux kernel extensions from external tree..
	#16093: kmod and gcc-14 build error

2024.02.2, released May 7th, 2024

	Important / security related fixes.

        Security: The /dev/shm entry in the default /etc/fstab used
        for busybox and sysv init systems was missing the sticky bit,
        allowing users to delete and replace other users files.

        Defconfigs: Mangopi mq1rdw1: Lock kernel headers to 6.6.x.

	Updated/fixed packages: apache, botan,
	brcmfmac_sdio-firmware-rpi, clamav, cmake, containerd,
	crucible, dmenu-wayland, dmraid, docker-cli, docker-engine,
	domoticz, fmc, freeipmi, gdal, gdb, giflib, gnupg2, go,
	google-breakpad, googlefontdirectory, imx-gpu-viv, inih,
	libassuan, libblockdev, libcurl, libdrm, libgpg-error,
	libksba, libnpth, libressl, libtracefs, lxc, mbedtls, mesa3d,
	oatpp, opensc, oprofile, perl-net-ssleay, petitboot,
	php-xdebug, psplash, python-pypng, python-segno,
	python-service-identity, python3, skeleton-init-sysv, systemd,
	tcf-agent, tiff, uuu, wolfssl, xmrig, zic

	Issues resolved (http://bugs.uclibc.org):

        #14416: Mesa3D missing DRI3 support for various Gallium drivers
        #16009: 'select' does not work with 'choice'
        #16036: android-tools version is too old

2024.02.1, released March 26th, 2024

	Important / security related fixes.

	Toolchain: Dropped old Codesourcery ARM / AARCH64 and
	Codescape MIPS external toolchains as they use old Glibc
	versions not compatible with 64-bit time_t

	Defconfigs: Starfive visionfive2: Use correct genimage
	configuration. TI am62x_sk / am64_sk: Use correct optee-os
	platform setting and use a fixed ti-k3-r5-loader version.

	Infrastructure: Use git to fetch CVE data from the FKIE
	repository to work around an issue with missing meta files in
	the releases.

	Updated/fixed packages: axel, bash, bitcoin, botan, busybox,
	containerd, cryptsetup, cvs, docker-compose, docker-engine,
	ell, expat, giflib, gmp, go, harfbuzz, intel-gmmlib,
	intel-mediadriver, iwd, kodi, kodi-pvr-hts, libcgroup, libdrm,
	libgit2, libkrb5, libnfs, libunwind, libvpl,
	libtorrent-rasterbar, libxml2, linux, mariadb, mesa3d,
	mmc-utils, msmtp, multipath-tools, neon, onevpl-intel-gpu,
	openssh, openvpn, osm2pgsql, owfs, php, poco, policycoreutils,
	postgis, privoxy, pulseaudio, pure-ftpd, putty,
	python-configshell-fb, python-rtslib-fb, rauc, restorecond,
	spice, squid, targetcli-fb, tzdata, util-linux, vdr, xmrig,
	xvisor, zic

	Issues resolved (http://bugs.uclibc.org):

	#15997: Build Squid with GCC12.3 in Buildroot 2024.02 will fail

2024.02, released March 5th, 2024

	Various fixes.

	Updated/fixed packages: botan, conmon, dash, efivar, expat,
	fluent-bit, gcc, gst1-vaapi, libcoap, libselinux, liburing,
	libva, libxml2, ltp-testsuite, luvi, newlib-bare-metal,
	nginx-naxsi, octave, openblas, openvmtools, parted, poco,
	powertop, python-aiohttp, python-cheroot, python-django,

shell/frameworks/namcap/namcap.py
#!/usr/bin/env python3
# Copyright (C) 2003-2023 Namcap contributors, see AUTHORS for details.
# SPDX-License-Identifier: GPL-2.0-or-later

import argparse
import os
import sys
import tarfile

import Namcap.depends
from Namcap.package import load_from_tarball, PacmanPackage
import Namcap.rules
import Namcap.tags
import Namcap.version


# Functions
def get_modules():
    """Return all possible modules (rules)"""
    return Namcap.rules.all_rules


def get_enabled_modules():
    """Return modules (rules) that should be used by default"""
    return dict(filter(lambda x: x[1].enable, get_modules().items()))


def open_package(filename):
    try:
        tar = tarfile.open(filename, "r")
        if ".PKGINFO" not in tar.getnames():
            tar.close()
            return None
    except IOError:
        if tar:
            tar.close()
        return None
    return tar


def show_messages(name, key, messages):
    colored_key = {
        "E": "\033[91mE\033[00m",
        "W": "\033[93mW\033[00m",
        "I": "\033[92mI\033[00m",
    }
    for msg in messages:
        if sys.stdout.isatty():
            print("%s %s: %s" % (name, colored_key[key], Namcap.tags.format_message(msg)))
        else:
            print("%s %s: %s" % (name, key, Namcap.tags.format_message(msg)))


def process_realpackage(package, modules):
    """Runs namcap checks over a package tarball"""
    pkgtar = open_package(package)

    if not pkgtar:
        print("Error: %s is empty or is not a valid package" % package)
        return 1

    pkginfo: PacmanPackage | None = load_from_tarball(package)
    if pkginfo is None:
        print(f"Error: Loading package from {package} failed")
        return 1

    # Loop through each one, load them apply if possible
    for i in modules:
        rule = get_modules()[i]()

        if isinstance(rule, Namcap.ruleclass.PkgInfoRule):
            rule.analyze(pkginfo, None)
        elif isinstance(rule, Namcap.ruleclass.PkgbuildRule):
            pass
        elif isinstance(rule, Namcap.ruleclass.TarballRule):
            rule.analyze(pkginfo, pkgtar)
        else:
            show_messages(pkginfo["name"], "E", [("error-running-rule %s", i)])

        # Output the three types of messages
        show_messages(pkginfo["name"], "E", rule.errors)
        show_messages(pkginfo["name"], "W", rule.warnings)
        if info_reporting:
            show_messages(pkginfo["name"], "I", rule.infos)

    # dependency analysis
    errs, warns, infos = Namcap.depends.analyze_depends(pkginfo)
    show_messages(pkginfo["name"], "E", errs)
    show_messages(pkginfo["name"], "W", warns)
    if info_reporting:
        show_messages(pkginfo["name"], "I", infos)


def process_pkginfo(pkginfo, modules):
    """Runs namcap checks of a single, non-split PacmanPackage object"""
    for i in modules:
        rule = get_modules()[i]()
        if isinstance(rule, Namcap.ruleclass.PkgInfoRule):
            rule.analyze(pkginfo, None)

        # Output the messages
        if "base" in pkginfo:
            name = "PKGBUILD (" + pkginfo["base"] + ")"
        else:
            name = "PKGBUILD (" + pkginfo["name"] + ")"
        show_messages(name, "E", rule.errors)
        show_messages(name, "W", rule.warnings)
        if info_reporting:
            show_messages(name, "I", rule.infos)


def process_pkgbuild(package, modules):
    """Runs namcap checks over a PKGBUILD"""
    # We might want to do some verifying in here... but really... isn't that
    # what pacman.load is for?
    pkginfo = Namcap.package.load_from_pkgbuild(package)

    if pkginfo is None:
        print("Error: %s is not a valid PKGBUILD" % package)
        return 1

    # apply global PKGBUILD rules
    for i in modules:
        rule = get_modules()[i]()
        if isinstance(rule, Namcap.ruleclass.PkgbuildRule):
            rule.analyze(pkginfo, package)
        # Output the messages
        if "base" in pkginfo:
            name = "PKGBUILD (" + pkginfo["base"] + ")"
        else:
            name = "PKGBUILD (" + pkginfo["name"] + ")"
        show_messages(name, "E", rule.errors)
        show_messages(name, "W", rule.warnings)
        if info_reporting:
            show_messages(name, "I", rule.infos)
    # apply per pkginfo rule
    for subpkg in pkginfo.subpackages if pkginfo.is_split else [pkginfo]:
        process_pkginfo(subpkg, modules)


# Main
modules = get_modules()
# Let's handle those options!
version = Namcap.version.get_version()

parser = argparse.ArgumentParser()
parser.add_argument("-L", "--list", action="store_true", help="List available rules")
parser.add_argument("-i", "--info", action="store_true", help="Prints information (debug) responses from rules")
parser.add_argument(
    "-m", "--machine-readable", action="store_true", help="Makes the output parseable (machine-readable)"
)
parser.add_argument("-t", "--tags", action="store", help="Use a custom tag file")
parser.add_argument("packages", nargs="*")
pargroup = parser.add_mutually_exclusive_group()
pargroup.add_argument(
    "-e",
    "--exclude",
    action="store",
    metavar="RULELIST",
    help="Don't apply RULELIST rules to the package (comma-separated)",
)
pargroup.add_argument(
    "-r",
    "--rules",
    action="store",
    metavar="RULELIST",
    help="Only apply RULELIST rules to the package (comma-separated)",
)
parser.add_argument("-v", "--version", action="version", version=version)
args = parser.parse_args()

if args.list:
    print("-" * 20 + " Namcap rule list " + "-" * 20)
    print(modules)
    for j in sorted(modules):
        print("%-20s: %s" % (j, modules[j].description))
    parser.exit(0)

if len(args.packages) == 0:
    print("Missing required argument packages", file=sys.stderr)
    parser.exit(2)

info_reporting = args.info
machine_readable = args.machine_readable
filename = args.tags
packages = args.packages

active_modules = {}

if args.rules:
    for rule in args.rules.split(","):
        if rule in modules:
            active_modules[rule] = modules[rule]
        else:
            print(f"Error: Rule '{rule}' does not exist")
            parser.exit(2)

if args.exclude:
    for rule in args.exclude.split(","):
        active_modules.update(modules)
        if rule in modules:
            active_modules.pop(rule)
        else:
            print(f"Error: Rule '{rule}' does not exist")
            parser.exit(2)

Namcap.tags.load_tags(filename=filename, machine=machine_readable)

# No rules selected?  Then use default selection
if len(active_modules) == 0:
    active_modules = get_enabled_modules()

# Go through each package, get the info, and apply the rules
for package in packages:
    if not os.access(package, os.R_OK):
        print("Error: Problem reading %s" % package)
        parser.print_usage()

    if os.path.isfile(package) and tarfile.is_tarfile(package):
        process_realpackage(package, active_modules)
    elif "PKGBUILD" in package:
        process_pkgbuild(package, active_modules)
    else:
        print("Error: %s not package or PKGBUILD" % package)
